lib/vrl/stdlib/src/hmac.rs

```diff
@@ -0,0 +1,243 @@
+use ::value::Value;
+use base64::Engine;
+use hmac::{Hmac as HmacHasher, Mac};
+use sha1::Sha1;
+use sha_2::{Sha224, Sha256, Sha384, Sha512};
+use vrl::prelude::*;
+
+macro_rules! hmac {
+    ($algorithm:ty, $key:expr, $val:expr) => {{
+        let mut mac =
+            <HmacHasher<$algorithm>>::new_from_slice($key.as_ref()).expect("key is bytes");
+        mac.update($val.as_ref());
+        let result = mac.finalize();
+        let code_bytes = result.into_bytes();
+        code_bytes.to_vec()
+    }};
+}
+
+fn hmac(value: Value, key: Value, algorithm: Value, encoding: Value) -> Resolved {
+    let value = value.try_bytes()?;
+    let key = key.try_bytes()?;
+    let algorithm = algorithm.try_bytes_utf8_lossy()?.as_ref().to_uppercase();
+    let encoding = encoding.try_bytes_utf8_lossy()?.as_ref().to_lowercase();
+
+    let code_bytes = match algorithm.as_str() {
+        "SHA1" => hmac!(Sha1, key, value),
+        "SHA-224" => hmac!(Sha224, key, value),
+        "SHA-256" => hmac!(Sha256, key, value),
+        "SHA-384" => hmac!(Sha384, key, value),
+        "SHA-512" => hmac!(Sha512, key, value),
+        _ => return Err(format!("Invalid algorithm: {}", algorithm).into()),
+    };
```

Comment:
Could this table be stored in a look up table that could also generate the `valid_algorithms` used below?

Reply:
No idea. Not something I want to take on at this point. I had enough fun making the compiler and type system happy by changing to a macro.

Reply:
Fair enough.